cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql

```diff
@@ -0,0 +1,135 @@
+ * @id cpp/unsafe-create-process-call
+ * @kind problem
+ * @problem.severity error
+ * @precision medium
+ * @tags security
+ *       external/cwe/cwe-428
+ *       external/microsoft/C6277
+ */
+ 
+import cpp
+import semmle.code.cpp.dataflow.DataFlow
+import semmle.code.cpp.dataflow.DataFlow2
+
+/**
+ * A function call to CreateProcess (either wide-char or single byte string versions)
+ */
+class CreateProcessFunctionCall extends FunctionCall {
+  CreateProcessFunctionCall() {
+    (
+      this.getTarget().hasGlobalName("CreateProcessA") or
+      this.getTarget().hasGlobalName("CreateProcessW") or
+      this.getTarget().hasGlobalName("CreateProcessWithTokenW") or
+      this.getTarget().hasGlobalName("CreateProcessWithLogonW") or
+      this.getTarget().hasGlobalName("CreateProcessAsUserA") or
+      this.getTarget().hasGlobalName("CreateProcessAsUserW") 
+    )
+  }
+  
+  int getApplicationNameArgumentId() {
+    if(
+      this.getTarget().hasGlobalName("CreateProcessA") or
+      this.getTarget().hasGlobalName("CreateProcessW") 
+    ) then ( result = 0 )
+    else if (
+      this.getTarget().hasGlobalName("CreateProcessWithTokenW") 
+    ) then ( result = 2 )
+    else if (
+      this.getTarget().hasGlobalName("CreateProcessWithLogonW") 
+    ) then ( result = 4 )
+    else if(
+      this.getTarget().hasGlobalName("CreateProcessAsUserA") or
+      this.getTarget().hasGlobalName("CreateProcessAsUserW") 
+    ) then ( result = 1 )
+    else (result = -1 )
+  }
+  
+  int getCommandLineArgumentId() {
+    if(
+      this.getTarget().hasGlobalName("CreateProcessA") or
+      this.getTarget().hasGlobalName("CreateProcessW") 
+    ) then ( result = 1 )
+    else if (
+      this.getTarget().hasGlobalName("CreateProcessWithTokenW") 
+    ) then ( result = 3 )
+    else if (
+      this.getTarget().hasGlobalName("CreateProcessWithLogonW") 
+    ) then ( result = 5 )
+    else if(
+      this.getTarget().hasGlobalName("CreateProcessAsUserA") or
+      this.getTarget().hasGlobalName("CreateProcessAsUserW") 
+    ) then ( result = 2 )
+    else (result = -1 )
+  }
+}
+ 
+/**
+ * Dataflow that detects a call to CreateProcess with a NULL value for lpApplicationName argument
+ */
+class NullAppNameCreateProcessFunctionConfiguration extends DataFlow::Configuration {
+  NullAppNameCreateProcessFunctionConfiguration() {
+    this = "NullAppNameCreateProcessFunctionConfiguration"
+  }
+
+  override predicate isSource(DataFlow::Node source) {
+    nullValue(source.asExpr()) 
+  }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists( 
+      CreateProcessFunctionCall call, Expr val |
+      val = sink.asExpr() |
+      val = call.getArgument(call.getApplicationNameArgumentId()) 
+    )
+  }
+}
+
+/**
+ * Dataflow that detects a call to CreateProcess with an unquoted commandLine argument
+ */
+class QuotedCommandInCreateProcessFunctionConfiguration extends DataFlow2::Configuration {
```

Comment:
I think this class does the opposite of what it's name implies.  `UnquotedCommandInCreateProcessFunctionConfiguration`?

Reply:
Good Catch. I started writing it the other way around, and I forgot to change the name. I will fix it